modules/example/src/smithy4s/example/GetObjectInput.scala

```diff
@@ -6,15 +6,21 @@ import smithy4s.ShapeId
 import smithy4s.ShapeTag
 import smithy4s.schema.Schema.struct
 
+/** Input for getting an Object
+  * all fields are required
+  * and are given through HTTP labels
+  */
 case class GetObjectInput(key: ObjectKey, bucketName: BucketName)
 object GetObjectInput extends ShapeTag.Companion[GetObjectInput] {
   val id: ShapeId = ShapeId("smithy4s.example", "GetObjectInput")
 
-  val hints: Hints = Hints.empty
+  val hints: Hints = Hints(
+    smithy.api.Documentation("Input for getting an Object\nall fields are required\nand are given through HTTP labels\nSee https://smithy.io/2.0/spec/http-bindings.html?highlight=httppayload#http-uri-label"),
+  )
 
   implicit val schema: Schema[GetObjectInput] = struct(
-    ObjectKey.schema.required[GetObjectInput]("key", _.key).addHints(smithy.api.HttpLabel(), smithy.api.Required()),
-    BucketName.schema.required[GetObjectInput]("bucketName", _.bucketName).addHints(smithy.api.HttpLabel(), smithy.api.Required()),
+    ObjectKey.schema.required[GetObjectInput]("key", _.key).addHints(smithy.api.Required(), smithy.api.Documentation("Sent in the URI label named \"key\".\nKey can also be seen as the filename\nIt is always required for a GET operation"), smithy.api.HttpLabel()),
+    BucketName.schema.required[GetObjectInput]("bucketName", _.bucketName).addHints(smithy.api.Required(), smithy.api.Documentation("Sent in the URI label named \"bucketName\"."), smithy.api.HttpLabel()),
```

Comment:
This doesn't seem to get rendered as a scaladoc

Reply:
that's the `aggregating the member documentations` part, right? I don't see this having been implemented yet.

Reply:
ah yes this should be rendered as `@constructor` arguments to `GetObjectInput` I believe, and I have not started yet with rendering those type of things

Reply:
Yup, so the correct way to go at it would be for `Hint.Documentation` type to change to : 

``` 
case class Documentation(docString: String, memberDocs: ListMap[String, String]) extends Hint
```

so that you'd be capturing the documentation of the members during the creation of the intermediate representation, and using it where it makes sense (ie, documentation on structures for now)

Reply:
Thank you for the hint! I understand that we want to capture the doc of the members (if there is any) in the IR. But how do I fill `memberDocs` in the `SmithyToIR.scala` file?

```scala
    case doc: DocumentationTrait =>
      Hint.Documentation(doc.getValue(), )
```

I've looked around a bit, and I cannot see where `DocumentationTrait` provides a function to get the member shapes. Do I have to make that logic myself?

Reply:
you wouldn't be using `DocumentationTrait` but <checks code> you'd need access to the shape whose hint you're rendering.

For example:

```diff
diff --git a/modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala b/modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala
index d2ac2bc2..affdff51 100644
--- a/modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala
+++ b/modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala
@@ -105,7 +105,7 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
     new ShapeVisitor[Option[Decl]] {
 
       private def getDefault(shape: Shape): Option[Decl] = {
-        val hints = traitsToHints(shape.getAllTraits().asScala.values.toList)
+        val hints = shapeHints(shape)
 
         val recursive = hints.exists {
           case Hint.Trait => true
@@ -218,7 +218,7 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
       }
 
       override def structureShape(shape: StructureShape): Option[Decl] = {
-        val hints = traitsToHints(shape.getAllTraits().asScala.values.toList)
+        val hints = shapeHints(shape)
         val isTrait = hints.exists {
           case Hint.Trait => true
           case _          => false
@@ -251,7 +251,7 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
       override def unionShape(shape: UnionShape): Option[Decl] = {
         val rec = isRecursive(shape.getId())
 
-        val hints = traitsToHints(shape.getAllTraits().asScala.values.toList)
+        val hints = shapeHints(shape)
         val isTrait = hints.exists {
           case Hint.Trait => true
           case _          => false
@@ -681,8 +681,8 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
 
     }
 
-  private def hints(shape: Shape): List[Hint] = traitsToHints(
-    shape.getAllTraits().asScala.values.toList
+  private def hints(shape: Shape): List[Hint] = shapeHints(
+    shape
   )
 
   def toTypeRef(id: ToShapeId): Type.Ref = {
@@ -730,7 +730,7 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
   @annotation.nowarn(
     "msg=class UniqueItemsTrait in package traits is deprecated"
   )
-  private val traitToHint: PartialFunction[Trait, Hint] = {
+  private def traitToHint(shape: Shape): PartialFunction[Trait, Hint] = {
     case _: ErrorTrait => Hint.Error
     case t: ProtocolDefinitionTrait =>
       val shapeIds = t.getTraits()
@@ -757,7 +757,9 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
     case ConstraintTrait(tr) => Hint.Constraint(toTypeRef(tr), unfoldTrait(tr))
   }
 
-  private def traitsToHints(traits: List[Trait]): List[Hint] = {
+  private def shapeHints(shape: Shape): List[Hint] = {
+    val traits = shape.getAllTraits().asScala.values.toList
+
     val nonMetaTraits =
       traits
         .filterNot(_.toShapeId().getNamespace() == "smithy4s.meta")
@@ -770,7 +772,9 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
     val nonConstraintNonMetaTraits = nonMetaTraits.collect {
       case t if ConstraintTrait.unapply(t).isEmpty => t
     }
-    traits.collect(traitToHint) ++ nonConstraintNonMetaTraits.map(unfoldTrait)
+    traits.collect(traitToHint(shape)) ++ nonConstraintNonMetaTraits.map(
+      unfoldTrait
+    )
   }
 
   case class AltInfo(name: String, tpe: Type, isAdtMember: Boolean)

```

then you can find the member docs like this (roughly):

```scala
val yourShapeMemberDocs = shape
  .members()
  .asScala
  .map(_.getTarget())
  .map(model.expectShape(_))
  .map(_.getTrait(classOf[DocumentationTrait]))
```

Reply:
that `private def hints(shape: Shape)` is kinda useless now. I only left it in because it has way too many usages... I guess you could remove it, rename `shapeHints` to `hints` and call it as `this.hints` wherever `hints` is taken.

Reply:
Ah I see! I'll get to this weekend! :D, that clears a few things up.
After that I'll work on the escaping of docstrings.
And hopefully then I can undraft this PR.

Reply:
sounds great. No pressure ;)

Reply:
> rename shapeHints to hints and call it as this.hints wherever hints is taken.

Do I rename it `this.hints`? The function `hints isn't part of `ShapeVisitor` but `SmithyIR`
So (from other functions) this seems to work:

```
      private def getDefault(shape: Shape): Option[Decl] = {
        val _hints = hints(shape)

```

Reply:
if I'm understanding it right, and `ShapeVisitor` is nested in `SmithyIR`, you could still call it as `SmithyIR.this.hints`... but I'm kinda too lazy to try it out in this given I'd have to repeat that diff above :joy:

just push what you have, we can iterate on that :)